From e70052dbb79411c36cc1be0ae3f71a09f7355915 Mon Sep 17 00:00:00 2001
From: Karl Bonde Torp <k.torp@samsung.com>
Date: Tue, 19 Sep 2023 10:08:29 +0200
Subject: [PATCH 1/4] lib/bdev: Add spdk_bdev_is_kv() to bdev interface

This patch is needed to enable KV passthru on fabrics, specifically to
set the correct csi when the namespace is added to nvmf/subsystem.

spdk_bdev_is_kv() will allow user to check if a
bdev is a kv bdev.

Signed-off-by: Karl Bonde Torp <k.torp@samsung.com>
---
 include/spdk/bdev.h        | 8 ++++++++
 include/spdk/bdev_module.h | 5 +++++
 lib/bdev/bdev.c            | 6 ++++++
 lib/nvmf/ctrlr.c           | 3 +++
 4 files changed, 22 insertions(+)

diff --git a/include/spdk/bdev.h b/include/spdk/bdev.h
index aa28b505d..6544171df 100644
--- a/include/spdk/bdev.h
+++ b/include/spdk/bdev.h
@@ -746,6 +746,14 @@ bool spdk_bdev_is_md_interleaved(const struct spdk_bdev *bdev);
  */
 bool spdk_bdev_is_md_separate(const struct spdk_bdev *bdev);
 
+/**
+ * Checks if bdev supports KV namespace semantics.
+ *
+ * \param bdev Block device to query.
+ * \return true if device supports KV namespace semantics.
+ */
+bool spdk_bdev_is_kv(const struct spdk_bdev *bdev);
+
 /**
  * Checks if bdev supports zoned namespace semantics.
  *
diff --git a/include/spdk/bdev_module.h b/include/spdk/bdev_module.h
index 4ca3feeb9..8ffdd3b80 100644
--- a/include/spdk/bdev_module.h
+++ b/include/spdk/bdev_module.h
@@ -563,6 +563,11 @@ struct spdk_bdev {
 	 */
 	uint32_t dif_check_flags;
 
+	/**
+	 * Specify whether bdev is kv device.
+	 */
+	bool kv;
+
 	/**
 	 * Specify whether bdev is zoned device.
 	 */
diff --git a/lib/bdev/bdev.c b/lib/bdev/bdev.c
index 88d01d4ee..cecc7a74e 100644
--- a/lib/bdev/bdev.c
+++ b/lib/bdev/bdev.c
@@ -4809,6 +4809,12 @@ spdk_bdev_is_md_separate(const struct spdk_bdev *bdev)
 	return (bdev->md_len != 0) && !bdev->md_interleave;
 }
 
+bool
+spdk_bdev_is_kv(const struct spdk_bdev *bdev)
+{
+	return bdev->kv;
+}
+
 bool
 spdk_bdev_is_zoned(const struct spdk_bdev *bdev)
 {
diff --git a/lib/nvmf/ctrlr.c b/lib/nvmf/ctrlr.c
index e78cf1a30..c5adba1bc 100644
--- a/lib/nvmf/ctrlr.c
+++ b/lib/nvmf/ctrlr.c
@@ -3188,6 +3188,9 @@ nvmf_ctrlr_identify_iocs(struct spdk_nvmf_ctrlr *ctrlr,
 		if (spdk_bdev_is_zoned(ns->bdev)) {
 			vector->zns = 1;
 		}
+		if (spdk_bdev_is_kv(ns->bdev)){
+			vector->kv = 1;
+		}
 	}
 
 	rsp->status.sct = SPDK_NVME_SCT_GENERIC;
-- 
2.44.0

